### ISSUE TITLE
Enable clang-tidy on all samples

### ISSUE THREAD

**Issue by torbsorb** [zivid/cpp-extra-samples#91]
You should not exclude samples here, then we lose out on testing of many files. Tooling should omit diagnostics for the headers of system libraries, so all additional diagnosticas triggered by this proposal will be in our files and should be fixed.

_Originally posted by @nedrebo in https://github.com/zivid/cpp-extra-samples/pull/85_

**Comment by torbsorb** [comment]
@nedrebo,

We do not install Eigen, PCL nor OpenCV in our CI. That's why we don't run build them, and why I believe they're not linted.

When we run clang-tidy it uses `CMakeLists.txt` which will check whether (e.g.) Eigen is installed. Is there an acceptable way to skip this check when `CMakeLists.txt` is invoke with clang-tidy?

I tried the following and it works:

in lint.sh:
```bash
...
cmake -GNinja \
    -DLINT_ONLY=ON \
    -DUSE_EIGEN3=ON \
...
```

in CMakeLists.txt:
```bash
...
if(USE_EIGEN3)
    if(LINT_ONLY)
        message(Will lint Eigen3 samples, but do not require Eigen3 installed)
    elseif(NOT DEFINED EIGEN3_INCLUDE_DIR)
...
```

Is this acceptable, or is there a better way.

Btw, I find that cmake remembers a variable set through -D if not set in a consecutive run. Maybe we should clear cmake-cache between runs? Or, is that some of what allows cmake to only rerun on changes?

**Comment by torbsorb** [comment]
I found -U :) So, we could just add:
```bash
...
cmake --build . || exit $?

cmake -ULINT_ONLY "$SOURCE_DIR" || exit $?
```

**Comment by nedrebo** [comment]
I am not sure I understand correctly what you ask for, but you must install all headers to be able to lint a. Code that cannot compile successfully cannot be analyzed.

**Comment by torbsorb** [comment]
It sort of worked, although you're correct that it will throw an error on missing header. Were able to fix most other issues though. See https://github.com/zivid/cpp-extra-samples/pull/92.

Maybe this PR will have to wait until we've fixed the CI so that it installs these dependencies.

**Comment by nedrebo** [comment]
Yes, I think you should first make all dependencies available in CI, then you don't need to do the weird workaround.

**Comment by SatjaSivcev** [comment]
This is done. clang-tidy now runs over all samples in CI — `lint.sh` configures the build with `-DUSE_PCL=ON -DUSE_EIGEN3=ON -DUSE_OPENCV=ON -DUSE_ARUCO=ON -DUSE_GENTL=ON` and all the optional dependencies are installed in the lint job, so every sample is linted. Closing.
